### PR TITLE
Support withdrawal/settlement

### LINKED ISSUE #43: Add support for withdraw processes
- [x] Support start payment challenge
 - [x] Support settlement after challenge period ended
 - [x] Support withdraw

### PR DESCRIPTION
**Pull request by katat**
Closes #43

### CODE REVIEW

File: src/commands/stage.js [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+                for (let ongoingChallenge of ongoingChallenges) {
+                    const {type, expirationTime, intendedStageAmount} = ongoingChallenge;
+                    const {amount} = intendedStageAmount.toJSON();
+                    const formattedStageAmount = ethers.utils.formatUnits(amount, tokenInfo.decimals);
+                    spinner.info(`Settlement type: ${type}; Stage amount: ${formattedStageAmount}; Expiration time: ${moment(expirationTime).format('dddd, MMMM Do YYYY, h:mm:ss a')}`);
+                }
+                return;
+            }
+            
+            let totalIntendedStageAmount = txs.reduce((cum, tx) => {
```
[the remark this answers is not in the document]
Comment (katat) [reply src/commands/stage.js]: Bingo :)
I will update this name to `accumulator`

**Comment by katat** [comment]
Sure, I have added the documentation for the new commands.

File: src/commands/settle.js [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+            const balanceBN = ethers.utils.parseUnits(balance, tokenInfo.decimals);
+
+            if (balanceBN.lt(stageAmountBN)) {
+                spinner.fail(`The maximum settleable nahmii balance is ${balance}`);
+                return;
+            }
+
+            spinner = ora('Calculating the required settlement challenges for the intended stage amount').start();
+            const {requiredChallenges, invalidReasons} = await settlement.getRequiredChallengesForIntendedStageAmount(stageMonetaryAmount, wallet.address);
```
Comment (hubiierik) [inline src/commands/settle.js]: Call is missing latestReceipt argument.

Comment (hubiierik) [inline src/commands/settle.js]: Return value (of promise) is an array. It cannot be unpacked.

File: src/commands/settle.js [hunk 3]
```diff
@@ -0,0 +1,85 @@
+'use strict';
+
+const dbg = require('../dbg');
+const nahmii = require('nahmii-sdk');
+const ethers = require('ethers');
+const moment = require('moment');
+const ora = require('ora');
+
+module.exports = {
+    command: 'settle <stageAmount> <currency> [--gas=<gasLimit> --price=<gasPrice in gwei>]',
```
Comment (morfj) [inline src/commands/settle.js]: stageAmount => amount?

**Review by morfj — CHANGES_REQUESTED** [review]
Should we also add an "unstage" command, or should that functionality be included in the deposit command?

**Comment by katat** [comment]
@morfj Thanks for the comments. I think it is a good idea to add the "unstage" command. I will make the updates according to the comments.

**Comment by morfj** [comment]
I think this PR could very well be labeled as v1.0.0 of the CLI :+1:

File: src/commands/settle.js [hunk 4]
```diff
@@ -26,21 +26,22 @@ module.exports = {
         yargs.coerce('amount', arg => arg);
     },
     handler: async (argv) => {
-        const { amount, currency, gas, price } = argv;
+        const { currency } = argv;
         const config = require('../config');
+        
         const provider = await nahmii.NahmiiProvider.from(config.apiRoot, config.appId, config.appSecret);
+        const tokenInfo = await provider.getTokenInfo(currency);
```
Comment (morfj) [inline src/commands/settle.js]: This async call is now done outside of the try...catch scope => unhandled rejection if it fails

File: src/commands/stage.js [hunk 5]
```diff
@@ -25,27 +25,28 @@ module.exports = {
         });
     },
     handler: async (argv) => {
-        const { currency, gas, price } = argv;
+        const { currency } = argv;
 
         const config = require('../config');
         const provider = await nahmii.NahmiiProvider.from(config.apiRoot, config.appId, config.appSecret);
+        const tokenInfo = await provider.getTokenInfo(currency);
```
Comment (morfj) [inline src/commands/stage.js]: Also outside of try...catch scope

File: src/commands/stage.js [hunk 6]
```diff
@@ -97,3 +98,10 @@ module.exports = {
         }
     }
 };
+
+function validatePositiveInteger(str) {
```
Comment (morfj) [inline src/commands/stage.js]: We could move this common function to the utils module I think.

File: src/commands/unstage.js [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+            desc: 'Gas price used',
+            default: 1,
+            type: 'number'
+        });
+    },
+    handler: async (argv) => {
+        const config = require('../config');
+
+        const provider = await nahmii.NahmiiProvider.from(config.apiRoot, config.appId, config.appSecret);
+        const tokenInfo = await provider.getTokenInfo(argv.currency);
```
Comment (morfj) [inline src/commands/unstage.js]: async outside try...catch

File: src/commands/stage.js (hunk 6 above)
Comment (morfj) [reply src/commands/stage.js]: Great minds think alike.. 🤣

**Review by morfj — COMMENTED** [review]
Maybe add a paragraph in the README about how you are supposed to use the settle, stage, withdraw commands in conjunction with each other, and indicate the "typical time" in between invocations. It would help explain the whole settlement process for the uninitiated I think.

**Review by morfj — APPROVED** [review]

**Pull request merged**